### ISSUE TITLE
Network actor blocks Client actor

### ISSUE THREAD

**Issue by ilblackdragon** [nearprotocol/nearcore#2463]
**Describe the bug**
We have multiple occasions we have observed network actor blocking client actor, which in turn fails to produce blocks.

Simplest way I observed what when you start a node, stats log that suppose to print every 10 seconds gets shown way less frequently:
```
Apr 14 07:32:22.240  INFO stats: Server listening at ed25519:4M6YwXQTAj5buGho96sYSvz1P9VYdE9UcHxqGBnmGsG3@0.0.0.0:24567
Apr 14 07:33:29.283  INFO stats: # 2451518 Downloading headers 98% V/8  4/4/40 peers  0 B/s  0 B/s 0.00 bps 0 gas/s CPU: 107%, Mem: 65.7 MiB
Apr 14 07:34:34.047  INFO stats: # 2451518 Downloading headers 98% V/8  5/5/40 peers  9.2kiB/s  27 B/s 0.00 bps 0 gas/s CPU: 100%, Mem: 70.8 MiB
Apr 14 07:35:40.295  INFO stats: # 2451518 Downloading headers 99% V/8  6/6/40 peers  7.1kiB/s  0.6kiB/s 0.00 bps 0 gas/s CPU: 98%, Mem: 72.9 MiB
```
See that it prints stats barely once a minute. In my case I have peers getting banned / reconnected all the time due to #2456

Additionally @ailisp observed this under sufficient load.

**Expected behavior**
Client actor is operating independent of how busy network actor is.

**Comment by mfornet** [comment]
Right now we have five different types of actors in our code:
- ClientActor
- ViewClientActor
- PeerManagerActor
- Peer
- TelemetryActor

For every active connection we span a new `Peer` actor, so it is expected that there will be up to 40 peer actors. I'm not sure how actix runtime works, but if resources are split "uniformly" among all actors, then networking (`PeerManagerActor` + `Peer`s) will take 90% of all the resources (with 40 active peers) and 71% (with 10 active peers). 

So maybe we should limit the resources used on each `Peer` instance. I assume the most expensive computation executed by this actor is receiving/sending data from the network (but this needs to be confirmed).

**Comment by nearmax** [comment]
The reason is because single actix actor is a single thread so while `ClientActor` is processing `NetworkClientMessages` this loop is blocked: https://github.com/nearprotocol/nearcore/blob/0e6a750352ef4d20aa283244101db49b0ef17815/chain/client/src/client_actor.rs#L162 . I don’t think increasing the number of actix core or anything like that will help here, because AFAIK it the fact that each actix actor is one thread is its core property. In this specific scenario quick fix can be to use `run_interval` instead of `run_later`, because `run_later` does not guarantee that it is going to be executed at the specific moment of time. The long-term solution though is to start cutting ClientActor  into sub-actors.

**Comment by ilblackdragon** [comment]
@nearmax The logs was just an observed example of actor being blocked.
Originally, we were planning to split Client Actor into more when we have a need. May be there is need now - but we should carefully design how we want this to look like.

@SkidanovAlex you mentioned, you were looking at this as well?

**Comment by bowenwang1996** [comment]
I feel like refactoring client actor can help mitigate the problem, but doesn't fully solve it because they can still get blocked by network actors.

**Comment by nearmax** [comment]
We should be super careful about splitting client actor. Having all our stuff under one actor allowed us to avoid race conditions (which was a strategic design decision back in the days). We might resurface race conditions when cutting client actor into more pieces, and race conditions are really time-consuming to diagnose and fix. I suggest we postpone separation of the client actor into subactors as much as possible until our nearcore is very stable and has a solid test coverage, and only then we start separating it into subactors.

**Comment by ilblackdragon** [comment]
This https://github.com/nearprotocol/nearcore/pull/2869 should be addressing this problem. And there is separate work item to split Client and ViewClient into separate threads: https://github.com/nearprotocol/nearcore/issues/2751.
